src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/BaseForTextBasedPlanPrinters.java

```diff
@@ -0,0 +1,101 @@
+				if ( planNumber < numberOfSiblings-1 ) {
+					return indentLevelString + nonLastChildIndentBase;
+				}
+				else {
+					return indentLevelString + lastChildIndentBase;
+				}
+			}
+			else if ( upperRootOpIndentString.endsWith(nonLastChildIndentBase) ) {
+				for ( int i = 1; i < planLevel; i++ ) {
+					indentLevelString += levelIndentBase;
+				}
+				if ( planNumber < numberOfSiblings-1 ) {
+					return indentLevelString + nonLastChildIndentBase;
+				}
+				else {
+					return indentLevelString + lastChildIndentBase;
+				}
+			}
+			else if ( upperRootOpIndentString.endsWith(lastChildIndentBase) ) {
+				indentLevelString = upperRootOpIndentString.substring( 0, upperRootOpIndentString.length() - lastChildIndentBase.length() ) + spaceBase;
+				if ( planNumber < numberOfSiblings-1 ) {
+					return indentLevelString + nonLastChildIndentBase;
+				}
+				else {
+					return indentLevelString + lastChildIndentBase;
+				}
+			}
+			else {
+				return indentLevelString;
+			}
+		}
+	}
+	
+	protected String getIndentLevelStringForDetail(final int planNumber, final int planLevel, final int numberOfSiblings, final int numberOfSubPlans, final String indentLevelString) {
+		String indentLevelStringForDetail = "";
+		if ( planLevel == 0 ) {
+			if ( numberOfSubPlans > 0 ) {
+				return "";
+			}
+			else {
+				return spaceBase;
+			}
+		}
+		if ( indentLevelString == "") {
+			indentLevelStringForDetail += spaceBase;
+		}
+		else if ( indentLevelString.endsWith(nonLastChildIndentBase) ) {
+			indentLevelStringForDetail = indentLevelString.substring( 0, indentLevelString.length() - nonLastChildIndentBase.length() ) + levelIndentBase;
+		}
+		else if ( indentLevelString.endsWith(lastChildIndentBase) && indentLevelString.startsWith(" ") ) {
+			indentLevelStringForDetail = indentLevelString.replaceAll( ".", " " );
+		}
+		else if ( indentLevelString.endsWith(lastChildIndentBase) && indentLevelString.startsWith(levelIndentBase) ) {
+			indentLevelStringForDetail = indentLevelString.substring( 0, indentLevelString.length() - lastChildIndentBase.length() ) + spaceBase;
+		}
+		else if ( indentLevelString.equals(lastChildIndentBase) ) {
+			indentLevelStringForDetail = indentLevelString.replaceAll( ".", " " );
+		}
+		return indentLevelStringForDetail;
+	}
+	
+	protected void printFederationMember( final FederationMember fm, final String indentLevelStringForOpDetail, final PrintStream out ) {
+		out.append( indentLevelStringForOpDetail + "  - fm (" + fm.getInterface().getID() + ") " + fm.getInterface().toString() );
+		out.append( System.lineSeparator() );
+	}
+	
+	protected void printSPARQLGraphPattern (final SPARQLGraphPattern gp, final String indentLevelStringForOpDetail, final PrintStream out ) {
+		out.append( indentLevelStringForOpDetail + "  - pattern (" + gp.hashCode() +  ") (" + gp.toString() + ")" );
+		out.append( System.lineSeparator() );
+		out.append( indentLevelStringForOpDetail );
+		out.append( System.lineSeparator() );
```

Comment:
Why these two lines? It may not always be the case that the `- pattern` line is the last of the detailed information lines. Whatever function uses this function here is the one that can decide whether an empty line should be printed below the `- pattern` line.

Reply:
> Do you mean, let the function calling printSPARQLGraphPattern decide whether to print an extra line?

Yes, exactly.